sound/soc/sof/ipc4-mtrace.c

```diff
@@ -0,0 +1,607 @@
+	kt = ktime_add_us(ktime_get_real(), FW_EPOCH_DELTA * USEC_PER_SEC);
+	system_time = ktime_to_us(kt);
+	msg.data_size = sizeof(system_time);
+	msg.data_ptr = &system_time;
+	if (iops->set_get_data(sdev, &msg, msg.data_size, true)) {
+		sdev->fw_trace_is_supported = false;
+		return 0;
+	}
+
+	msg.extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(SOF_IPC4_FW_PARAM_ENABLE_LOGS);
+
+	priv->state_info.enable = 1;
+
+	msg.data_size = sizeof(priv->state_info);
+	msg.data_ptr = &priv->state_info;
+
+	priv->mtrace_state = SOF_MTRACE_INITIALIZING;
+	if (iops->set_get_data(sdev, &msg, msg.data_size, true)) {
+		sdev->fw_trace_is_supported = false;
+		priv->mtrace_state = SOF_MTRACE_DISABLED;
+		return 0;
+	}
+
+	priv->mtrace_state = SOF_MTRACE_ENABLED;
+
+	return 0;
+}
+
+static void ipc4_mtrace_disable(struct snd_sof_dev *sdev)
+{
+	struct sof_mtrace_priv *priv = sdev->fw_trace_data;
+	const struct sof_ipc_ops *iops = sdev->ipc->ops;
+	struct sof_ipc4_msg msg;
+	int i;
+
+	if (priv->mtrace_state == SOF_MTRACE_DISABLED)
+		return;
+
+	msg.primary = SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG);
+	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	msg.primary |= SOF_IPC4_MOD_ID(SOF_IPC4_MOD_INIT_BASEFW_MOD_ID);
+	msg.primary |= SOF_IPC4_MOD_INSTANCE(SOF_IPC4_MOD_INIT_BASEFW_INSTANCE_ID);
+	msg.extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(SOF_IPC4_FW_PARAM_ENABLE_LOGS);
+
+	priv->state_info.enable = 0;
+
+	msg.data_size = sizeof(priv->state_info);
+	msg.data_ptr = &priv->state_info;
+	iops->set_get_data(sdev, &msg, msg.data_size, true);
+
+	priv->mtrace_state = SOF_MTRACE_DISABLED;
+
+	for (i = 0; i < sdev->num_cores; i++) {
+		struct sof_mtrace_core_data *core_data = &priv->cores[i];
+
+		core_data->host_read_ptr = 0;
+		core_data->dsp_write_ptr = 0;
+		wake_up(&core_data->trace_sleep);
+	}
+}
+
+/*
+ * Parse the slot descriptors at debug_box offset, we are only interested in the
+ * type of the slot to map it to a core
+ */
+static void sof_mtrace_find_core_slots(struct snd_sof_dev *sdev)
+{
+	struct sof_mtrace_priv *priv = sdev->fw_trace_data;
+	struct sof_mtrace_core_data *core_data;
+	u32 slot_desc_offset, type, core;
+	int i;
+
+	/*
+	 * Out of the 16 (MAX_MTRACE_SLOTS) slots the first one is allocated for
+	 * descriptors to describe the remaining 15 slots (MAX_MTRACE_SLOTS - 1)
+	 * We are checking the descriptors to find the slot <-> core mapping
+	 */
+	slot_desc_offset = sdev->debug_box.offset;
+	for (i = 0; i < MAX_MTRACE_SLOTS - 1; i++) {
+		/* Read the slot type */
+		sof_mailbox_read(sdev, slot_desc_offset + 4, &type, 4);
+		if ((type & SOF_MTRACE_SLOT_TYPE_MASK) == SOF_MTRACE_SLOT_DEBUG_LOG) {
+			core = type & SOF_MTRACE_SLOT_CORE_MASK;
+
+			if (core >= sdev->num_cores)
+				continue;
+
+			core_data = &priv->cores[core];
+			core_data->slot_offset = sdev->debug_box.offset;
+			core_data->slot_offset += SOF_MTRACE_SLOT_SIZE * (i + 1);
+			dev_dbg(sdev->dev, "%s: slot%d is used for core%u\n",
+				__func__, i, core);
```

Comment:
should this be i+1, if slot 0 is used for descriptors it cannot be used for a specific core, can it?

It would be simpler to start the loop at 1 really!
for (i = 1; i < MAX_MTRACE_SLOTS; i++) {

Reply:
> should this be i+1, if slot 0 is used for descriptors it cannot be used for a specific core, can it?

The first usable slot as a real slot is the second, that is true. The debug window is split up to equal sized chunks, mostly referred as slots. The first one is the descriptor are. In a way it is a slot but from another angle it is not really a slot as it can not be used for anything else than for the descriptors.
The first (index 0) descriptor at the start of the debug window is describing the slot at debug_window + slot_size, since the descriptor are is made to be the same size as the other slots.

> It would be simpler to start the loop at 1 really! for (i = 1; i < MAX_MTRACE_SLOTS; i++) {

Depending on how we are looking at this, yes, but for me, when I look at for (i = 1; ...) my brain will say first that oh, we are skipping the first entry in this loop. But we don't. We read the 0th descriptor as well, which describes the first slot _after_ the descriptor area.